cmd/ceremony/main.go

```diff
@@ -0,0 +1,280 @@
+			return errors.New("if key-type = \"rsa\" then ecdsa-curve is not used")
+		}
+		if cc.KeyType == "ecdsa" && cc.ECDSACurve == "" {
+			return errors.New("if key-type = \"ecdsa\" then ecdsa-curve is required")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.IssuerPath != "" {
+			return errors.New("issuer-path is not used for key ceremonies")
+		}
+		if cc.CertificatePath != "" {
+			return errors.New("certificate-path is not used for key ceremonies")
+		}
+		if cc.CertificateProfile != nil {
+			return errors.New("certificate-profile is not used for key ceremonies")
+		}
+	}
+
+	return nil
+}
+
+func main() {
+	configPath := flag.String("config", "", "Path to ceremony configuration file")
+	flag.Parse()
+
+	if *configPath == "" {
+		log.Fatal("--config is required")
+	}
+	configBytes, err := ioutil.ReadFile(*configPath)
+	if err != nil {
+		log.Fatalf("Failed to read config file: %s", err)
+	}
+	var config ceremonyConfig
+	err = json.Unmarshal(configBytes, &config)
+	if err != nil {
+		log.Fatalf("Failed to parse config: %s", err)
+	}
+
+	if err = config.Validate(); err != nil {
+		log.Fatalf("Failed to validate config: %s", err)
+	}
+
+	ctx, session, err := pkcs11helpers.Initialize(config.PKCS11Module, config.KeySlot, "")
+	if err != nil {
+		log.Fatalf("Failed to setup session and PKCS#11 context for slot %d: %s", config.KeySlot, err)
+	}
+	log.Printf("Opened PKCS#11 session for slot %d\n", config.KeySlot)
+
+	switch config.CeremonyType {
+	case "root":
+		keyInfo, err := generateKey(ctx, session, config)
+		if err != nil {
+			log.Fatal(err)
+		}
+		signer, err := getKey(ctx, session, config.KeyLabel, keyInfo.id)
+		if err != nil {
+			log.Fatalf("Failed to retrieve signer: %s", err)
+		}
+		template, err := makeTemplate(ctx, session, config.CertificateProfile, keyInfo.der)
+		if err != nil {
+			log.Fatalf("Failed to create certificate profile: %s", err)
+		}
+		// x509.CreateCertificate uses a io.Reader here for signing methods that require
+		// a source of randomness. Since PKCS#11 based signing generates needed randomness
+		// at the HSM we don't need to pass a real reader. Instead of passing a nil reader
+		// we use one that always returns errors in case the internal usage of this reader
+		// changes.
+		certBytes, err := x509.CreateCertificate(&failReader{}, template, template, keyInfo.key, signer)
+		if err != nil {
+			log.Fatalf("Failed to create certificate: %s", err)
+		}
+		log.Printf("Signed certificate: %x\n", certBytes)
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			log.Fatalf("Failed to parse signed certificate: %s", err)
+		}
+		if err := cert.CheckSignatureFrom(cert); err != nil {
+			log.Fatalf("Failed to verify certificate signature: %s", err)
+		}
+		log.Println("Verified certificate signature")
+		pemBytes := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certBytes})
+		log.Printf("Certificate PEM:\n%s", pemBytes)
```

Comment:
Let's log the PEM bytes before we check the signature, so if we do fail that check we can look at the bytes and figure out what went wrong.